### PR TITLE
Fully deprecate Python 2

### PR DESCRIPTION
**Pull request by kiwiz**
- [x] Remove Python 2 references
- [x] Integrate `ruff`
- [x] Migrate to Github Actions
- [x] Fix errors with Sphinx docs
- [x] Review Sphinx docs
- [ ] Testing

Backwards incompatible changes:

- Bump minimum Python version to 3.10
- `Keep.login`/`Keep.resume`/`APIAuth.login` no longer return a `bool` 
- The `device_id` parameter previously defaulted to `get_mac()`, which returned an `int`. This has been updated to cast the int to a hexadecimal string

### CODE REVIEW

File: gkeepapi/__init__.py [hunk 1]
```diff
@@ -171,20 +168,20 @@ def refresh(self):
...
 
 
-class API(object):
+class API:
     """Base API wrapper"""
 
     RETRY_CNT = 2
 
-    def __init__(self, base_url, auth=None):
+    def __init__(self, base_url: str, auth: APIAuth = None):
```
Comment (wizpig64) [inline gkeepapi/__init__.py:184]: Apparently the current kosher in type hinting is to explicitly use `Optional`  (or ` | None` in python 3.10) if any param defaults to `None`:

```diff
-    def __init__(self, base_url: str, auth: APIAuth = None):
+    def __init__(self, base_url: str, auth: Optional[APIAuth] = None):

```

https://adamj.eu/tech/2022/10/18/python-type-hints-implicit-optional-types/

**Comment by kiwiz** [comment]
Let's just support Python 3.10+ for all the nice new type annotation features :firecracker:

**Comment by wizpig64** [comment]
There are some [hacks](https://pypi.org/project/future-typing/) to get the 3.10 syntax working as far back as 3.6, and 3.9 doesn't go EOL until 2025. But if you want to keep the version list short for simplicity, I'd support that decision.

**Comment by kiwiz** [comment]
That's a neat (ab?)use of codecs. I'll stick with 3.10 for now and see if anyone has issues.

**Pull request merged**